doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,894 @@
+`reveal_locals()` can also be placed on any line and will tell you the inferred types of all local variables.
+
+```python
+from typing import Union, List
+
+
+def print_things(thing: Union[int, List[int]]) -> None:
+    if isinstance(thing, int):
+        reveal_type(thing)
+        print(thing)
+
+    reveal_type(thing)
+    print(st for st in thing)
+```
+
+Running `mypy` on the above reveals the types that `mypy` sees:
+
+```cmd
+main.py:5: note: Revealed type is "builtins.int"
+main.py:8: note: Revealed type is "Union[builtins.int, builtins.list[builtins.int]]"
+main.py:9: error: Item "int" of "Union[int, List[int]]" has no attribute "__iter__" (not iterable)
+Found 1 error in 1 file (checked 1 source file)
+```
+
+These debugging functions don't need to be imported from anywhere and are only recognized by `mypy` - therefore you will
+need to remove them from code before runtime.
+
+### Use typing.Any sparingly
+
+`typing.Any`, like the name suggests, assumes that a type can be _anything_ - it is compatible with every other type,
+and vice versa. This is what allows gradual typing in Python, as an untyped function like this...
+
+```python
+def add_things(x, y):
+    return x + y
+```
+
+is seen by the type checker as this...
+
+```python
+from typing import Any
+
+
+def add_things(x: Any, y: Any) -> Any:
+    return x + y
+```
+
+`Any` should be used sparingly since it effectively turns the type checker off. If the type can be narrowed down to
+something more specific, that is preferred since over usage of `Any` can undermine the type checker's ability to find
+bugs in other parts of the code.
+
+### Use typing.Union when accepting more than one type
+
+[typing.Union](https://docs.python.org/3/library/typing.html#typing.Union) allows for typing an argument with more than
+one type.
+
+```python
+from typing import Union
+
+
+def foo(bat: Union[float, str]) -> None:
+    ...
+```
+
+Tips:
+A `Union` requires at least two types and is more useful with types that are not consistent with each other. For
+example, usage of `Union[int, float]` is not necessary since `int` is consistent with `float` -- just use `float`. It's
+also recommended trying to avoid having functions return `Union` types as it causes the user to need to understand/parse
+through the return value before acting on it. Sometimes this can be resolved by using an [overload](TODO).
+
+### Use typing.Optional when a parameter default is None
+
+Python's typing docs for [Optional](https://docs.python.org/3/library/typing.html#typing.Optional) explain that
+
+> Optional[X] is equivalent to X | None (or Union[X, None])
+
+This is somewhat confusingly named. This means that `Optional` should only be used when an argument can be either
+something _or_ `None`. For example, `Optional` usage is appropriate here:
+
+```python
+from typing import Optional
+
+
+def tree(kind: Optional[str] = None):
+    ...
+```
+
+But not necessary here since a default argument of "oak" is provided:
```

Comment:
Well, technically if `kind` can be `None`, then it's `Optional`. It never says that the default value _must_ be `None`

```python
def tree(kind: Optional[str]):  # I can be None, I have no default value, I'm required to be specified
    ...

def tree(kind: Optional[str] = "oak"):  # By default use "oak", but setting `None` is acceptable
    ...

```

Yes, this is terrible naming, and for those previous scenarios I think using `Union[X, None]` is more clear actually.

Reply:
Good point, added this to the doc!

Reply:
Good point, added this to the doc!